whisper/src/net.rs

```diff
@@ -0,0 +1,638 @@
+
+use message::{Message, Error as MessageError};
+
+const RALLY_TOKEN: TimerToken = 1;
+const RALLY_TIMEOUT_MS: u64 = 750; // supposed to be at least once per second.
+
+const PROTOCOL_VERSION: usize = 2;
+
+/// Supported protocol versions.
+pub const SUPPORTED_VERSIONS: &'static [u8] = &[PROTOCOL_VERSION as u8];
+
+// maximum tolerated delay between messages packets.
+const MAX_TOLERATED_DELAY_MS: u64 = 2000;
+
+/// Number of packets.
+pub const PACKET_COUNT: u8 = 3;
+
+mod packet {
+	pub const STATUS: u8 = 0;
+	pub const MESSAGES: u8 = 1;
+	pub const TOPIC_FILTER: u8 = 2;
+}
+
+/// Handles messages within a single packet.
+pub trait MessageHandler: Send + Sync {
+	/// Evaluate the message and handle it.
+	///
+	/// The same message will not be passed twice.
+	/// Heavy handling should be done asynchronously.
+	/// If there is a significant overhead in this thread, then an attacker
+	/// can determine which kinds of messages we are listening for.
+	fn handle_messages(&self, message: &[Message]);
+}
+
+// errors in importing a whisper message.
+#[derive(Debug)]
+enum Error {
+	Decoder(DecoderError),
+	Network(NetworkError),
+	Message(MessageError),
+	UnknownPacket(u8),
+	UnknownPeer(PeerId),
+	ProtocolVersionMismatch(usize),
+	SameNodeKey,
+	UnexpectedMessage,
+}
+
+impl From<DecoderError> for Error {
+	fn from(err: DecoderError) -> Self {
+		Error::Decoder(err)
+	}
+}
+
+impl From<NetworkError> for Error {
+	fn from(err: NetworkError) -> Self {
+		Error::Network(err)
+	}
+}
+
+impl From<MessageError> for Error {
+	fn from(err: MessageError) -> Self {
+		Error::Message(err)
+	}
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		match *self {
+			Error::Decoder(ref err) => write!(f, "Failed to decode packet: {}", err),
+			Error::Network(ref err) => write!(f, "Network error: {}", err),
+			Error::Message(ref err) => write!(f, "Error decoding message: {}", err),
+			Error::UnknownPacket(ref id) => write!(f, "Unknown packet kind: {}", id),
+			Error::UnknownPeer(ref id) => write!(f, "Message received from unknown peer: {}", id),
+			Error::ProtocolVersionMismatch(ref proto) =>
+				write!(f, "Unknown protocol version: {}", proto),
+			Error::UnexpectedMessage => write!(f, "Unexpected message."),
+			Error::SameNodeKey => write!(f, "Peer and us have same node key."),
+		}
+	}
+}
+
+// sorts by work proved, descending.
+#[derive(PartialEq, Eq)]
+struct SortedEntry {
+	slab_id: usize,
+	work_proved: OrderedFloat<f64>,
+	expiry: SystemTime,
+}
+
+impl Ord for SortedEntry {
+	fn cmp(&self, other: &SortedEntry) -> Ordering {
+		self.work_proved.cmp(&other.work_proved)
+	}
+}
+
+impl PartialOrd for SortedEntry {
+	fn partial_cmp(&self, other: &SortedEntry) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+// stores messages by two metrics: expiry and PoW rating
+// when full, will accept messages above the minimum stored.
+struct Messages {
+	slab: ::slab::Slab<Message>,
+	sorted: Vec<SortedEntry>,
```

Comment:
a binary heap will be optimized for the use case when we tend to pop off the first element. we tend to remove elements by two metrics: when they expire (~random access) and also by pushing out low PoW elements (sorted access). My intuition was that vector pops (removing low work entries) are very cheap `O(n)` worst case, and a vector `retain` periodically to prune expired messages is also relatively cheap `O(n)`. 

Doing an arbitrary number of pops from a binary heap to remove low work entries is `O(nlogn)`, and the retainment logic for live messages would have to reconstruct the heap each time from the iterator, again `O(nlogn)`